### PR TITLE
Make Tree Table Sortable

### LINKED ISSUE #4811: Allow sorting of Tree like model list views
### Environment
* Nautobot version (Docker tag too if applicable): 2.0
* Python version: 3.11
* Database platform, version: psql
* Middleware(s): N/A

### Steps to Reproduce
1. Go to https://demo.nautobot.com/ipam/prefixes/
2. Click on headers
3.

### Expected Behavior

To sort based on the column header you created

### Observed Behavior

No sorting

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes #4811 
# What's Changed

# Screenshots

### Before Sorting

Screenshot 2024-02-02 at 2 47 30 PM

### After Sorting

Screenshot 2024-02-02 at 2 47 16 PM

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/tables.py [hunk 1]
```diff
@@ -146,6 +143,38 @@ def configurable_columns(self):
...
+
+    @order_by.setter
+    def order_by(self, value):
+        """
+        Order the rows of the table based on columns.
+
+        Arguments:
+            value: iterable or comma separated string of order by aliases.
+        """
+        # NOTE: Copy paste code from parent order_by; since calling `super().order_by` raises an Exception
```
Comment (glennmatthews) [inline nautobot/core/tables.py:158]: Do we know why it would raise an Exception?

File: nautobot/core/tables.py [hunk 2]
```diff
@@ -146,6 +143,38 @@ def configurable_columns(self):
...
+            name = OrderBy(alias).bare
+            if name in self.columns and self.columns[name].orderable:
+                valid.append(alias)
+        self._order_by = OrderByTuple(valid)
+
+        model = getattr(self.Meta, "model", None)
+        if model and issubclass(model, TreeNode):
+            self.data.data = self.data.data.extra(order_by=self.order_by)
+        else:
+            self.data.order_by(self._order_by)
```
Comment (glennmatthews) [inline nautobot/core/tables.py:176]: Can you add a comment explaining what part of this differs from the base class order_by behavior? I assume it's just the TreeNode check and case?

File: nautobot/core/tests/test_tables.py [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
 class TableTestCase(TestCase):
-    def test_model_table_orderable(self):
-        """Assert TreeNode model table orderable is set to False, and non tree model table is True by default"""
-        tree_node_model_tables = (LocationTypeTable, LocationTable, RackGroupTable)
-        non_tree_node_model_tables = (DeviceTypeTable, ManufacturerTable)
+    def _validate_sorted_tree_queryset_same_with_table_queryset(self, queryset, table_class, field_name):
+        with self.subTest(f"Assert sorting {table_class.__name__} on '{field_name}'"):
+            table = table_class(queryset, order_by=field_name)
+            table_queryset_data = table.data.data.values_list("pk", flat=True)
+            sorted_queryset = queryset.with_tree_fields().extra(order_by=[field_name]).values_list("pk", flat=True)
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/tests/test_tables.py:13]: Yeah it sorts as expected 100% sure, but which test can i add to assert this?

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/tests/test_tables.py:13]: Was thinking maybe doing an actual python `sort()` call on one or the other of the lists, but if you're confident that it's working as intended that may not be needed.

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/tests/test_tables.py:14]: Can I assume that since you marked this as resolved you're not planning to add the additional test?

**Review by glennmatthews — APPROVED** [review]

File: nautobot/core/tests/test_tables.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/tests/test_tables.py:14]: Yup 👍

**Pull request merged**